### PR TITLE
Support Stdout and Exception Details when Test Setup Fails

### PR DESCRIPTION
**Pull request by jtylwalk**
Adds support to `buck test` to capture stdout / report exceptions correctly when setup of test fails - if `@BeforeTest` or `@BeforeMethod` fails, any child test will also be marked as failing with the failure exception and output from the setup method. This aligns with the expected behavior of testng, where if setup fails, any child tests will be skipped.

@Addepar/infrastructure @davidhao3300

### CODE REVIEW

File: src/com/facebook/buck/testrunner/TestNGRunner.java [hunk 1]
```diff
@@ -44,8 +45,12 @@
  * Class that runs a set of TestNG tests and writes the results to a directory.
  */
 public final class TestNGRunner extends BaseRunner {
+
+  private static TestSuiteFailureListener suiteFailureListener;
+
   @Override
   public void run() throws Throwable {
+
```
Comment (davidhao3300) [inline src/com/facebook/buck/testrunner/TestNGRunner.java]: Unneeded newline?

**Comment by davidhao3300** [comment]
this is awesome, lgtm
should probably wait for an LGTM from @jiangty-addepar or @liuyang-li though

File: test/com/facebook/buck/testrunner/testdata/simple_testng_setup/test/SimpleTest.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2014-present Facebook, Inc.
```
Comment (bdworth) [inline test/com/facebook/buck/testrunner/testdata/simple_testng_setup/test/SimpleTest.java:2]: I don't know Addepar's procedures for using open-source code in our code base but thought I'd flag this as a possible Bad Thing (TM). Or maybe it's fine.

Comment (jtylwalk) [inline test/com/facebook/buck/testrunner/testdata/simple_testng_setup/test/SimpleTest.java:2]: All files _must_ contain this license preamble - it's actually checked within their codebase by one of their unit tests

**Comment by jtylwalk** [comment]
Slight warning about this - now that we're capturing more stdout, there's a more likely potential to run out of memory - see https://jenkinstest.addepar.com/job/AMP-unit-test-runner-buck/10269 for reference. @navkast previously encountered the same issue and bumped the memory for buck up in this commit cce5274199c3e1d2cb129422fc0ceed50d89fa89, but that wasn't ever merged to master.

Refactoring buck to stream stdout/stderr instead of hold in memory is a pretty large undertaking, so we may need to permanently bump up the allocated memory.

**Comment by jiangty-addepar** [comment]
looks exciting! @liuyang-li and I will take a look today, hopefully

File: src/com/facebook/buck/testrunner/TestNGRunner.java [hunk 3]
```diff
@@ -134,6 +142,76 @@ public void transform(ITestAnnotation annotation, Class testClass, Constructor t
...
+      originalErr = System.err;
+      rawStdOutBytes = new ByteArrayOutputStream();
+      rawStdErrBytes = new ByteArrayOutputStream();
+      stdOutStream = streamToPrintStream(rawStdOutBytes, System.out);
+      stdErrStream = streamToPrintStream(rawStdErrBytes, System.err);
+      System.setOut(stdOutStream);
+      System.setErr(stdErrStream);
+    }
+
+    public void onConfigurationFailure(ITestResult itr) {
```
Comment (jiangty-addepar) [inline src/com/facebook/buck/testrunner/TestNGRunner.java:173]: why no `@Override`?

**Comment by jtylwalk** [comment]
Comments addressed, can I get another review?

File: src/com/facebook/buck/testrunner/TestNGRunner.java [hunk 4]
```diff
@@ -134,6 +141,79 @@ public void transform(ITestAnnotation annotation, Class testClass, Constructor t
     }
   }
 
+  /**
+   * Listens for suite configuration / failure, captures stdout/stderr when configuring for
```
Comment (liuyang-li) [inline src/com/facebook/buck/testrunner/TestNGRunner.java:145]: nitpick: extra `/`?

File: src/com/facebook/buck/testrunner/TestNGRunner.java [hunk 5]
```diff
@@ -134,6 +141,79 @@ public void transform(ITestAnnotation annotation, Class testClass, Constructor t
...
+   * properly reporting both output and exceptions to child test cases
+   */
+  private static class TestSuiteFailureListener extends TestListenerAdapter {
+    private PrintStream originalOut, originalErr, stdOutStream, stdErrStream;
+    private ByteArrayOutputStream rawStdOutBytes, rawStdErrBytes;
+    private Throwable throwable;
+
+    //buffers for holding std out/err when configuring:
```
Comment (liuyang-li) [inline src/com/facebook/buck/testrunner/TestNGRunner.java:153]: nitpick: space after `//`.

**Comment by liuyang-li** [comment]
LGTMBC. Would you please run `ant clean && ant lint && ant` and make sure there is no errors before pushing?

File: src/com/facebook/buck/testrunner/TestNGRunner.java (hunk 4 above)
Comment (jtylwalk) [inline src/com/facebook/buck/testrunner/TestNGRunner.java:145]: Not sure what you mean - that comment is basically saying 'suite configuration or failure'

Comment (liuyang-li) [inline src/com/facebook/buck/testrunner/TestNGRunner.java:145]: I see. I saw in other places when `/` is used for or, there is not spaces between the words and the `/`. Either removing the spaces or changing to `or` would be fine I feel.

**Comment by davidhao3300** [comment]
Any updates on this? Would like to test whether JenkinsTest can handle more concurrent tests now that we've upped the heap limit

**Comment by jtylwalk** [comment]
Not too concerned with the nitpick comments, so this is basically ready to merge. I've already been using it on a branch so I know it's working.

**Comment by sreenath-are** [comment]
LGTM :shipit:

**Pull request merged**